frontend/ui/widget/inputdialog.lua

```diff
@@ -470,6 +472,29 @@ function InputDialog:init()
             },
         }
     end
+    if self._added_widgets then
+        for _, widget in ipairs(self._added_widgets) do
+            self:addWidget(widget, true)
+        end
+    end
+end
+
+function InputDialog:addWidget(widget, no_backup)
+    if not no_backup then -- backup widget for re-init
+        if not self._added_widgets then
+            self._added_widgets = {}
+        end
+        table.insert(self._added_widgets, widget)
+    end
+    local centered_widget = CenterContainer:new{
```

Comment:
Is this right to have any such widget generically centered ? Shouldn't the caller provide the thing CernterContainer'ed is that's what it wishes ?
Is it this centering that allowed you to get rid of checkbox_shift ?

Reply:
That works for your checkbox since you make them TextBoxWidget with a max_width, so they actually take this max_width as their width, right ?
With the old ones using a TextWidget, you could then have them misaligned, right?
```
      [ ] Case sensitive
  [ ] Use regular expression
```
Feels it would be then harder to counteract this imposed centering, no ?

Reply:
> With the old ones using a TextWidget

Do they exist?

Reply:
May be not yet, but think about the future :)
We're designing an API, so we want it to work in the most cases (current and future).
At the time your CheckBoxes were TextWidget without a fixed width, you would have been annoyed if there was this :addWidget() available, and it did this centering - and you may have had a hard time fighting the centering.
Just asking :) I'm ok with keeping it if it make the current use cases shorter - but I just want you to give it a thought too and not regret it later.

Reply:
I thought that centering of 4 only existing added widgets would be a good optimization.
Also I thought that cenetering would not prevent shifting of elements (if needed in the future) in the caller, as we know input_dialog.width and can calculate self width of the added widget's content.

Reply:
OK then.
(About optimisation, you could do the CenterContainer wrapping in the `if not no_backup` and store the CenterContainer itself in self._added_widgets - so you don't need to do it each time.)

Reply:
Trying to get it. If so, the caller would pass uncentered widget, :init would pass centered widget.
What would be the third argument in
https://github.com/koreader/koreader/blob/c27fc9c40bc60f1b654717ca646a7d7b2d25c467/frontend/ui/widget/inputdialog.lua#L497
Or to have two conditional table.insert's?

Reply:
```lua
function InputDialog:addWidget(widget, no_backup)
    if not no_backup then -- backup widget for re-init
        -- Center it for a better look
        widget = CenterContainer:new{
            dimen = Geom:new{
                w = self.width,
                h = widget:getSize().h,
            },
            widget,
        }
        if not self._added_widgets then
            self._added_widgets = {}
        end
        table.insert(self._added_widgets, widget)
    end
    -- insert widget before the bottom buttons and their previous vspan
    table.insert(self.vgroup, #self.vgroup-1, widget)
end
```
May be `re_init` or `re_add` would then sound better than `no_backup`.

Reply:
> ```lua
> widget = CenterContainer:new{
>             dimen = Geom:new{
>                 w = self.width,
>                 h = widget:getSize().h,
>             },
>             widget,
>         }
> ```

Nice